File: leet/1293_shortest_path_in_grid_with_obstacle_elimination/1293_shortest_path_in_grid_with_obstacle_elimination.py
# LeetCode 1293. Shortest Path in a Grid with Obstacles Elimination
# https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/
# O(n*m*k) time-and-space
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        m, n, steps = len(grid), len(grid[0]), 0
        dirs = ((0, 1), (1, 0), (-1, 0), (0, -1))
        visited = [[float('inf')] * n for _ in range(m)]
        queue = deque()
        queue.append([0, 0, 0])
        while queue:
            for _ in range(len(queue)):
                i, j, _k = queue.popleft()
                if i == m - 1 and j == n - 1:
                    return steps
                for ddi, ddj in dirs:
                    di, dj = i + ddi, j + ddj
                    if (0 <= di < m and 0 <= dj < n and
                        _k + grid[di][dj] <= k and
                        _k + grid[di][dj] < visited[di][dj]):
                            visited[di][dj] = _k + grid[di][dj]
                            queue.append([di, dj, _k + grid[di][dj]])
            steps += 1
        return -1

File: leet/1293_shortest_path_in_grid_with_obstacle_elimination/1293_shortest_path_in_grid_with_obstacle_elimination.ts
// LeetCode 1293. Shortest Path in a Grid with Obstacles Elimination
// https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/
// O(n*m*k) time-and-space
function shortestPath(grid: number[][], k: number): number {
    let m = grid.length, n = grid[0].length;
    if (k >= m + n - 2)
        return m + n - 2;
    const dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    const queue = [[0, 0, 0]];
    const visited = new Array(m).fill(0).map(a => new Array(n).fill(Number.MAX_SAFE_INTEGER));
    let x = 0, steps = 0;
    while (x < queue.length) {
        for (let sz = queue.length; x < sz; x++) {
            let [i, j, _k] = queue[x];
            if (i == m - 1 && j == n - 1)
                return steps;
            for (let [ddi, ddj] of dirs) {
                let di = i + ddi, dj = j + ddj;
                if (di >= 0 && di < m && dj >= 0 && dj < n) {
                    let nextK = grid[di][dj] + _k;
                    if (nextK <= k && visited[di][dj] > nextK) {
                        visited[di][dj] = nextK;
                        queue.push([di, dj, nextK]);
                    }
                }
            }
            
        }
        steps++;
    }
    return -1;
};
